# وعده‌های خود را بگیرید

> وقتی که جوان‌تر بودم، در گرفتن هر توپی بدترین بودم. بنابراین تصمیم گرفتم با همیشه گرفتن وعده‌هایم جبران کنم.

اشتباه شماره یک که برنامه‌نویسان جاوا اسکریپت بی‌تجربه مرتکب می‌شوند، نگرفتن وعده‌هایشان است.

## مثال بد

به مثال زیر توجه کنید:

```js
async function getSomething() {
  return fetch('https://something.com');
}

something.on('my_event', async () => {
  const something = await getSomething();
  console.log(something);
});
```

چه اتفاقی می‌افتد وقتی `getSomething()` رد شود؟ در Node.js، یک رویداد `unhandledRejection` رخ می‌دهد و فرآیند شما را می‌کشد! چون در طول آزمایش `something.com` آنلاین بود، ممکن است هرگز با این مشکل مواجه نشده باشید و فکر کنید کدتان عالی است. اما اینطور نیست، واقعاً بد است.

نمی‌توانید هر متدی را فقط `async` کنید! هر وعده _باید_ گرفته شود. این مثل Yin & Yang، نمک و فلفل و لیلی و مارشال است. آن‌ها به هم تعلق دارند.

## یک مثال کمی بهتر

```js
something.on('my_event', async () => {
  try {
    const something = await getSomething();
    console.log(something);
  } catch( err ) {
    console.error('Error getting something:', err);
  }
});
```

ببینید چطور حالا وعده را می‌گیریم؟ اما هنوز عالی نیست. چه اتفاقی می‌افتد وقتی کسی کدی بالاتر از `try {` اضافه کند؟ همچنان می‌تواند بدون گرفتن رد شود! احتمال آن کم است، اما ممکن است. پس بهتر است ایمن باشیم تا پشیمان.

## بهترین مثال

بنابراین، این شکل نهایی زیبایی ماست:

```js
something.on('my_event', () => {
  // از یک زمینه همزمان، یک زمینه غیرهمزمان جدید ایجاد کنید
  Promise.resolve().then(async () => {
    const something = await getSomething();
    console.log(something);
  }).catch(err => {
    console.error('Error getting something:', err);
  });
});
```

مدیر رویداد `my_event` همزمان است، همان‌طور که انتظار می‌رود، و _در داخل_ متد همزمان، یک زمینه _غیرهمزمان_ جدید با فراخوانی `Promise.resolve().then(async () => { ... })` ایجاد می‌کنیم و در نهایت با اضافه کردن `.catch(err => { ... })` آن را می‌گیریم.

این الگو شما را مجبور می‌کند به _اینکه چه باید کرد وقتی این وعده رد شد؟_ فکر کنید، به جای اینکه روز خود را ادامه دهید و اجازه دهید کسی با تجربه‌تر اشتباه شما را پاک کند.

## نکته اضافی: دو نوع زنجیره وعده

در تقریباً همه موارد، دو نوع زنجیره وعده وجود دارد.

**اولی یک زنجیره وعده است که از یک درخواست/پاسخ شروع می‌شود.** برای مثال، یک سرور وب که در حال رسیدگی به یک درخواست است، یا یک کاربر که دکمه‌ای را فشار می‌دهد.

این ساده‌ترین حالت برای کدنویسی است. فقط متد بالاترین سطح را غیرهمزمان کنید، یک مدیریت‌گر خطا پیوست کنید (مثلاً بازگشت HTTP 5XX و ارسال خطا به صورت JSON) و تمام کد خود را هنگام پایین آمدن انتظار بکشید.

**دومی یک زنجیره وعده است که به صورت خارجی شروع می‌شود.** برای مثال، از طریق یک ارسال‌کننده رویداد، یا یک مسیر موازی در کد شما که _داشتن آن خوب است_ (مثلاً، ارسال پیام به Slack هنگام اجرای یک کرون‌جاب).

در این مورد، فکر کنید رد شدن وعده چه باید کند (تقریباً همیشه: لاگ کردن در کنسول)، و سپس آن را _در یک زنجیره وعده جدید_ با استفاده از الگوی `Promise.resolve().then(async () => { ... }).catch(err => { ... })` کدنویسی کنید.